qr-code/src/main/java/com/adyen/checkout/qrcode/FullQRCodeView.kt

```diff
@@ -0,0 +1,212 @@
+        val resId = getMessageTextResource(paymentMethodType) ?: return
+        binding.textViewTopLabel.text = localizedContext.getString(resId)
+    }
+
+    private fun updateLogo(paymentMethodType: String?) {
+        Logger.d(TAG, "updateLogo - $paymentMethodType")
+        if (!paymentMethodType.isNullOrEmpty()) {
+            binding.imageViewLogo.loadLogo(
+                environment = delegate.componentParams.environment,
+                txVariant = paymentMethodType,
+                size = LogoSize.LARGE
+            )
+        }
+    }
+
+    private fun updateQrImage(qrCodeData: String?) {
+        Logger.d(TAG, "updateQrImage - $qrCodeData")
+        if (!qrCodeData.isNullOrEmpty()) {
+            val imagePath = QR_IMAGE_BASE_PATH + qrCodeData
+            Logger.d(TAG, "QR URL - $imagePath")
+            binding.imageViewQrcode.load(
+                environment = delegate.componentParams.environment,
+                path = imagePath
+            )
+        }
+    }
+
+    @StringRes
+    private fun getMessageTextResource(paymentMethodType: String?): Int? {
+        return when (paymentMethodType) {
+            PaymentMethodTypes.PAY_NOW -> R.string.checkout_qr_code_pay_now
+            else -> null
+        }
+    }
+
+    private fun onTimerTick(timerData: TimerData) {
+        val minutes = TimeUnit.MILLISECONDS.toMinutes(timerData.millisUntilFinished)
+        val seconds = TimeUnit.MILLISECONDS.toSeconds(timerData.millisUntilFinished) % TimeUnit.MINUTES.toSeconds(1)
+
+        val minutesSecondsString = localizedContext.getString(
+            R.string.checkout_qr_code_time_left_format,
+            minutes,
+            seconds
+        )
+
+        binding.textViewTimer.text = localizedContext.getString(
+            R.string.checkout_qr_code_pay_now_timer_text,
+            minutesSecondsString
+        )
+        binding.progressIndicator.progress = timerData.progress
+    }
+
+    @SuppressLint("MissingPermission")
+    private fun saveQrImage(qrCodeData: String?) {
+        val baseUrl = delegate.componentParams.environment.baseUrl
+        val imageUrl = baseUrl + QR_IMAGE_BASE_PATH + qrCodeData
+        Logger.d(TAG, "saveQrImage $imageUrl")
+        findViewTreeLifecycleOwner()?.lifecycleScope?.launchWhenCreated {
+            fileDownloader.download(
+                context,
+                imageUrl,
+                fileName,
+                android.os.Environment.DIRECTORY_DOWNLOADS,
+                MIME_TYPE
+            ) {
+                // TODO: add translation for downloading message
+                context.toast("qr code downloaded successfully")
+            }
+        }
+    }
+
+    override fun getView(): View = this
+
+    override fun highlightValidationErrors() = Unit
+
+    companion object {
+        private val TAG = LogUtil.getTag()
+        private const val QR_IMAGE_BASE_PATH = "barcode.shtml?barcodeType=qrCode&fileType=png&data="
```

Comment:
This REALLY should not be in the view, same as file name, mime type and `android.os.Environment.DIRECTORY_DOWNLOADS` 😬